### PR TITLE
first commit for weight implementation

### PR DESCRIPTION
**Pull request by valeriadamante**
this PR is created to integrate the weight computation in anatuples

### CODE REVIEW

File: pu.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        jsonFile = puWeightProducer.jsonPath.format(period)
+        if not puWeightProducer.initialized:
+            headers_dir = os.path.dirname(os.path.abspath(__file__))
+            header_path = os.path.join(headers_dir, "pu.h")
+            ROOT.gInterpreter.Declare(f'#include "{header_path}"')
+            ROOT.gInterpreter.ProcessLine(f'::correction::puCorrProvider::Initialize("{jsonFile}")')
+            puWeightProducer.initialized = True
+
+    def getWeight(self, df):
+        for scale in ['Central', 'Up', 'Down']:
```
Comment (kandrosov) [inline pu.py:19]:
```suggestion
        for scale in getScales():
```
and modify `getScales` in `CorrectionsCore.py`:
```python
def getScales(source=None):
    if source is None:
        return [ central, up, down ]
    if source == central:
        return [ central ]
    return [ up, down ]
```

File: pu.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        static const std::map<UncScale, std::string> names = {
+            { UncScale::Down, "down" },
+            { UncScale::Central, "nominal" },
+            { UncScale::Up, "up" },
+        };
+        return names.at(scale);
+    }
+    puCorrProvider(const std::string& fileName) :
+        corrections_(CorrectionSet::from_file(fileName)),
+        puweight(corrections_->at("Collisions18_UltraLegacy_goldenJSON"))
```
Comment (kandrosov) [inline pu.h:34]: this is era-dependant -> should be passed as an argument to the constructor and defined in a dedicated `dict` in `pu.py`.

File: tau.py [hunk 3]
```diff
@@ -30,3 +30,11 @@ def getES(self, df, source_dict):
...
 
         return df, source_dict
+
+
+    def getESWeight(self, df, weight_list):
+        for source in [ central ] + TauCorrProducer.energyScaleSources:
+            for scale in getScales(source):
+                syst_name = getSystName(source, scale)
+                updateWeightDict(weight_list, syst_name)
+        return df, weight_list
```
Comment (kandrosov) [inline tau.py:40]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Corrections.py [hunk 4]
```diff
@@ -59,4 +64,33 @@ def applyScaleUncertainties(df):
                 if obj not in source_objs:
                     suffix = 'Central' if obj in [ "Tau", "MET" ] else 'nano'
                     df = df.Define(f'{obj}_p4_{syst_name}', f'{obj}_p4_{suffix}')
-    return df, syst_dict
\ No newline at end of file
+    return df, syst_dict
+
+
+
+def getWeights(df, config=None, sample=None):
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Corrections.py:71]: ok

File: Corrections.py [hunk 5]
```diff
@@ -59,4 +64,33 @@ def applyScaleUncertainties(df):
...
+    lumi = config['GLOBAL']['luminosity']
+    sampleType = config[sample]['sampleType']
+    with open('config/crossSections.yaml', 'r') as xs_file:
+        xs_dict = yaml.safe_load(xs_file)
+    xs_name = config[sample]['crossSection']
+    xs = xs_dict[xs_name]['crossSec']
+    xs_stitching = 1
+    xs_stitching_incl = 1
+    xs_inclusive = 1
+    stitching_weight = 1
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Corrections.py]: ok

File: CorrectionsCore.py [hunk 6]
```diff
@@ -24,4 +24,8 @@ def updateSourceDict(source_dict, source, obj):
         source_dict[source] = []
     if obj in source_dict[source]:
         raise RuntimeError(f"addUncSource: dupblicated {source} definition for {obj}")
-    source_dict[source].append(obj)
\ No newline at end of file
+    source_dict[source].append(obj)
+
+def updateWeightDict(weight_list, source):
+    if source not in weight_list:
+        weight_list.append('w_'+source)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply CorrectionsCore.py:31]: ok

File: pu.py (hunk 1 above)
Comment (valeriadamante) [reply pu.py:21]: ok

File: tau.py (hunk 3 above)
Comment (valeriadamante) [reply tau.py]: ok

File: Corrections.py [hunk 7]
```diff
@@ -59,4 +64,33 @@ def applyScaleUncertainties(df):
...
+    if sampleType == 'DY' or sampleType=='W':
+        xs_stitching_name = config[sample]['crossSectionStitch']
+        inclusive_sample_name = 'DYJetsToLL_M-50' if sampleType=='DY' else 'WJetsToLNu'
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Corrections.py:85]: ok

File: corrections.h [hunk 8]
```diff
@@ -19,4 +19,18 @@ enum class UncScale : int {
     Up = +1,
 };
 
-}
\ No newline at end of file
+}
+
+float GetStitchingWeight(const SampleType& sampleType, const float LHE_Vpt, const int LHE_Njets ){
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply corrections.h:24]: ok

File: pu.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include "correction.h"
+#include "corrections.h"
+
+namespace correction {
+
+class puCorrProvider {
+public:
+
+    static void Initialize(const std::string& fileName)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply pu.h:10]: ok

**Pull request merged**